lib/fhir_models/validation/element_validator/terminology_validator.rb

```diff
@@ -0,0 +1,102 @@
+module FHIR
+  module Validation
+    # Validate Terminology in coded elements
+    class TerminologyValidator
+      attr_reader :vs_validators
+      def initialize(validators = {})
+        @vs_validators = validators
+      end
+
+      # Method for registering ValueSet and CodeSystem Validators
+      #
+      # @param valueset_uri [String] The ValueSet URI the validator is related to
+      # @param validator [#validate] The validator
+      def register_vs_validator(valueset_uri, validator)
+        @vs_validators[valueset_uri] = validator
+      end
+
+      # Clear the registered ValueSet and CodeSystem Validators
+      def clear_vs_validators
+        @vs_validators = {}
+      end
+
+      # Verify that the element meets terminology requirements
+      #
+      # @param element [Object] The Element of the Resource under test
+      # @param element_definition [FHIR::ElementDefinition] The Element Definition for the elements
+      # @return result [FHIR::ValidationResult] The result of the terminology check
+      def validate(resource, element_definition)
+        elements = FHIR::Validation::Retrieval.retrieve_by_element_definition(resource,
+                                                                              element_definition,
+                                                                              indexed: true)
+        elements.flat_map do |path, el|
+          validate_element(el, element_definition, path)
+        end
+      end
+
+      def validate_element(element, element_definition, path)
+        results = []
+        # Get the type
+        type_code = if element_definition.type.one?
+                      element_definition.type.first.code
+                    else
+                      return UnknownType.new('Need path in order to determine type') unless path
+
+                      element_definition.type.find do |datatype|
+                        cap_code = "#{datatype.code[0].capitalize}#{datatype.code[1..-1]}"
+                        /[^.]+$/.match(element_definition.path.gsub('[x]', cap_code)) == /[^.]+$/.match(path)
+                      end.code
+                    end
+
+        if %w[CodeableConcept Coding Quantity].include? type_code
```

Comment:
Ya ultimately is should support `code` and `uri` too.  I left these out because the existing validator relies on code and system to be present and I want to put more thought into how to support validating these elements which just have codes.

https://www.hl7.org/fhir/terminologies.html